### ISSUE TITLE
All those ignored NULL errors

### ISSUE THREAD

**Issue by Alcaro** [libretro/RetroArch#4497]
I see if (RANDOM_POINTER == NULL) return false; approximately everywhere. Most of them are for stuff that should never be null.

If those branches are taken, it's a bug. Ignoring bugs leads to strange, irreproducible issues and undebuggable walls of broken spaghetticode. While ignoring the error yields slightly better short-term results for the end user, the only good result is having the bug reported and fixed, which requires reporting it somewhere.

Luckily, there is a way to get both benefits: Some macros that prints something on stderr, calls a breakpointable function, and returns whatever failure value is appropriate. Something like [g_return_if_fail](https://developer.gnome.org/glib/stable/glib-Warnings-and-Assertions.html#g-return-if-fail), or maybe just if (!EXPECT_NOT_NULL(the_pointer)) return false; (but still a macro, we need \_\_FILE\_\_ and \_\_LINE\_\_). I don't care which, as long as we do something.

It's nearly impossible to find which NULL checks should be possible to hit, and which are guaranteed bugs. But can we at least stop adding more?

**Comment by twinaphex** [comment]
Fair enough, we can start looking into ways to mitigate and reduce the number of NULL early returns after 1.4.0. It will have to be well tested though.

**Comment by Alcaro** [comment]
I'm willing to implement the given macros, as long as we can reach an agreement on suitable names and behavior. To get the bikeshedding started, I propose the following:
```
int foo(int*x) {
 /* if the condition is false, prints a message and returns 42 */
 ASSERT_OR_RET(x, 42);
 return 7+*x;
}
void bar(int*x) {
 /* blank arguments are valid in macros; this expands to 'return ;'
    we could use a separate macro for void, but I can't think of a good name
    we should not use plain ASSERT, people expect that one to abort if false */
 ASSERT_OR_RET(x,);
 *x = 4;
}
int* baz(int*x) {
 int*y = malloc(sizeof(int));
 /* perhaps malloc failures should have a separate macro that silences future assertions?
    that way, we can stick assertions on 'this function only fails if malloc does'
      and have them not throw a stacktrace of cascading errors
    or maybe silence all assertions once ten have been fired, to not spam the log
    or maybe something even weirder
    more bikeshedding needed */
 ASSERT_OR_RET(y!=NULL, NULL);
 /* alternative macro, to be used if more cleanup than 'return x' is needed */
 if (!ASSERT_OR_FALSE(x!=NULL)) { free(y); return NULL; }
 *y = 9+*x;
 return y;
}
```
 

I believe removing spurious null checks is best done by liberally converting suspicious checks and seeing what happens. If they're non-spurious, the program is still behaving as expected, so the program will still work and it's easy to turn them back.

Many checks should be converted at once, so they can be tested simultaneously. If one turns out so noisy it drowns out the others, fix that one and repeat the test; this should be rare.

**Comment by twinaphex** [comment]
If we do this, will you be willing to do more work on the RetroArch codebase in general in the future? Previously you have been decidedly disinterested in it so I was wondering if this  would see some increased activity from you in return if we were to go through all this.

Also, I think performance is a factor here. Get rid of most of the messy null checks sure, but we need to be very careful with the functions that get called every frame. Function calls and runtime checks like this can affect performance negatively, and since we are now even targeting DOS/Windows 9x, performance is now more of a factor than ever. Codebase niceties should not come at the expense of performance at least.

**Comment by Alcaro** [comment]
Yes, I will make sure it does not slow anything down unless the assertions fail. In fact, I'll throw in the GCC hint for 'optimize assuming this is true' - it could actually end up faster!

The program size will grow a bit (not counting debug info, I'd expect around 20-30 bytes per assertion), but I'll add an ifdef we can flip on joke platforms. Just keep it enabled on modern desktops, okay?

The main reason I don't like working on RetroArch is because it's a huge mess of barely-documented interfaces with enormous amounts of criss-crossing dependencies. If I change something, I can't know what risks breaking; if something does break, I can't know which component is violating its specification; and I often can't even find the guilty component.

These spurious null checks contribute to that problem, by removing my ability to know out whether null is a valid argument to various functions.

(The other reason is I don't use RetroArch, so I don't know what needs to be done. But that one can be fixed by telling me what to do.)

**Comment by Alcaro** [comment]
oh, and this whole retro_load_game(NULL) nonsense is quite demotivating as well. If you're going to demand dozens of cores expect nulls there rather than tell frontend to not create nulls, what other bugs are you hacking around that I'm not aware of? What other ways are you trying to break our forward compatibility promises that I'm not aware of?

**Comment by twinaphex** [comment]
> What other ways are you trying to break our forward compatibility promises that I'm not aware of?

I think this is bordering on the paranoia. Exactly nothing is going on here, I have no intention break forward compatibility and I am one of the guys who insists highest on both backwards and forwards compatibility, to the point where I have routinely refused just bumping the API and trying to make the most of what we have with the current API. You can't find a person who would care about such things to this extent anywhere else I'd bet.

**Comment by twinaphex** [comment]
I suggest we start with this post-1.4.0 (which I hope I can push out of the door this Monday or so, presuming the testing goes well)
